[WIP] Expanded SF types to include MM types.

evetion opened the pull request:
Resolves #2 and resolves #3.

visr commented:
Nice! Would be nice to have focused PRs though, perhaps all those TODOs are a bit much for all in this PR?

evetion commented:
Sure. I won't add more implementations/easy checks here. But the Types are there now, including a bounding box one. Let me know what you think. Only thing left is maybe add a method bounds that results in a BoundingBox with a fallback that computes it from all coordinates.

visr commented on src/GeoInterfaceRFC.jl:
```diff
@@ -5,47 +5,84 @@ struct Geometry <: AbstractGeometry end
 
 # All Geometries
-function geomtype(x::Any)
+function geomtype(x::T) where T
```
Is there any reason you prefer `(x::T) where T` over `(x::Any)`? I suspect the `::Any` was there to make it extra explicit that it is a fallback, but we can also just leave it out `(x)`.

visr commented on test/runtests.jl:
```diff
@@ -0,0 +1,2 @@
+using GeoInterfaceRFC
+using Test
```
Could you add some tests for `BoundingBox`?

visr commented on src/GeoInterfaceRFC.jl:
```diff
@@ -5,47 +5,84 @@ struct Geometry <: AbstractGeometry end
+abstract type AbstractPolyHedralSurface <: AbstractSurface end
 struct Polygon <: AbstractPolygon end
+struct Triangle <: AbstractPolygon end  # Polygon with just three points
+struct Box <: AbstractPolygon end  # Polygon with just 4 points
```
Hmm this one is not in your picture at https://github.com/JuliaGeo/GeoInterfaceRFC.jl/issues/3. Do you think it's important to add it? And what about naming it `Rectangle`, since `Box` could be interpreted as a 3D structure?

visr replied:
I guess this is not necessarily orthogonal to the axis, like the `BoundingBox`, right?

asinghvi17 replied:
I guess this would be a `Quad` in the general sense (if it can have 4 points)?

evetion replied:
It's not a Quad, but it could be another Special type. The Box is rectangular, while the Quad is more generic and just specifies four corners and can thus also be diamond shaped.

evetion replied:
I added `Quad` as well.

evetion commented:
After some thought I removed the bounding box type. It had some issues, such as the `@generated` and it doesn't fit in this package as a concrete implementation. I would like to go for an optional `extent` method that returns something that adheres to a `Rectangle` type. The `Rect` type from GeometryBasics would be a good fit.

I've added the `CircularCurve` and related types from SQL/MM. Also added a lot of defaults, where we can infer that ourselves. I'm ok with `NotImplementedError`s on functions where we ourselves could eventually provide a fallback, things like `area`.

I've also started adding docstrings and did a split for types and interfaces, as requested by @rafaqz in #2.

Some points that still need to be addressed:
- Do we need to AbstractX everywhere? Is a `Triangle <: Polygon` or `<:AbstractPolygon`?
- I propose a minimal required interface, only `geomtype`, `ncoords`, `ngeom`, `getgeom`. We can make npoint = ngeom in the case of `Point`s etc. Even `Polygon`s, with exterior and interior(s) can be resolved to `nrings` (made an example) and thus `ngeoms`.
- `getgeom` can be an iterator.

Still some work to do, but let me know your thoughts on this direction.

visr commented on src/defaults.jl:
```diff
@@ -1,4 +1,24 @@
+isclosed(::AbstractMultiCurve, geom) = all([isclosed(getgeom(geom, i)) for i in 1:ngeom(geom)])
+
 
+issimple(::MultiPoint, geom) = allunique([getgeom(geom, i) for i in 1:ngeom(geom)])
```
`all` also supports a predicate function, such that you can use:
`all(i -> issimple(getgeom(geom, i)), 1:ngeom(geom))`

For `allunique` this is not defined, but changing the `[]` to `()` also speeds it up considerably, since no temporary array is allocated.

visr commented on src/defaults.jl:
```diff
@@ -0,0 +1,37 @@
+
+issimple(::AbstractCurve, geom) = allunique([getpoint(geom, i) for i in 1:npoint(geom) - 1]) && allunique([getpoint(geom, i) for i in 2:npoint(geom)])
+isclosed(::AbstractCurve, geom) = getpoint(geom, 1) == getpoint(geom, npoint(geom))
+isring(::AbstractCurve, geom) = issimple(::AbstractCurve, geom) && isclosed(::AbstractCurve, geom)
```
invalid syntax

visr commented on src/interface.jl:
```diff
@@ -0,0 +1,89 @@
+getpoint(geom, i::Integer) = getpoint(geomtype(geom), geom, i)
+
+# Curve
+startpoint(geom)
```
This and more below is a function call, not a function definition. If there is no good fallback definition, we can also just declare an empty function, `function startpoint end`.

visr commented on src/types.jl:
```diff
@@ -0,0 +1,68 @@
+"""A `LineString` is a collection of straight lines between its `Point`s."""
+struct LineString <: AbstractCurve end
+"""A Line is `LineString` with just two points."""
+struct Line <: LineString end
```
This is invalid, you can only subtype abstract structs. Should probably be `AbstractLineString`.

visr commented on Project.toml:
```diff
@@ -2,3 +2,6 @@ name = "GeoInterfaceRFC"
 authors = ["Yeesian Ng <ngyeesian@gmail.com>"]
 version = "0.1.0"
+
+[deps]
```
Since Test is only needed when running tests and not for the package, this should go under extras, like this:
```
[extras]
Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"

[targets]
test = ["Test"]
```

evetion replied:
Do we want 1.0, 1.1 compat? Otherwise we can go for a Project.toml in the tests folder.

visr replied:
I think we want the 1.0 compat if we can, I wouldn't drop it for this reason.

visr commented on src/GeoInterfaceRFC.jl:
```diff
@@ -5,4 +5,26 @@ include("interface.jl")
 include("defaults.jl")
 include("primitives.jl")
 
+export
+Point
+LineString
+Line
+LinearRing
+CircularString
+CompoundCurve
+CurvePolygon
+Polygon
+Triangle
+Rectangle
+Quad
+Pentagon
+Hexagon
+PolyHedralSurface
+TIN
+MultiPoint
+MultiCurve
+MultiLineString
+MultiPolygon
+GeometryCollection
```
I'm thinking, since these types are only used as traits and are not otherwise usable, maybe we should not export them? This will also clash with for instance several of the exports from GeometryBasics.

evetion replied:
That's a good point. I do assume we export all other functions? Like ngeom, ncoord, etc.

visr replied:
Not sure about 'all', but perhaps those that are expected to be used by end users rather than interface implementers?

evetion commented:
- Updated the exports and the `Project.toml`
- Updated the `README.md` to work with `ngeom` and `getgeom`
- Removed the `primitives.jl` for now (not included it), as it needs rethinking
- Made a test implementation as test
- Made a testing tool to check your implementation

I'd say, barring obvious errors or typos, it's ready to merge. It isn't done yet, as the test suite needs expanding, especially for the default cases, but it's good enough for JuliaCon.
